### PR TITLE
added static typechecking guide

### PR DESCRIPTION
**Pull request by ishankbahl**
Issue #76 

Added "Static Type Checking" guide.

### CODE REVIEW

File: content/docs/static-type-checking.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+As your app grows, you can catch and eliminate an entire category of bugs with static typechecking. You can use JavaScript extensions like [Flow](https://flowtype.org/) or [TypeScript](https://www.typescriptlang.org/) to statically typecheck your whole application. In complex applications, we recommend to use Flow or TypeScript for typechecking instead of typechecking with PropTypes.
+
+## Using Flow with React
+
+For trying out Flow, use this [playground](https://flow.org/try/).
+
+Flow and [Babel](http://babeljs.io/) work well together, so it doesn’t take much to adopt Flow as a React user who already uses Babel.
+
+### With babel
```
Comment (yangshun) [inline content/docs/static-type-checking.md]: Nit: Capital B in Babel

File: content/docs/static-type-checking.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+Flow and Babel are designed to work great together. It takes just a few steps to set them up together.
+
+If you don’t have Babel setup already, you can do that by following [this guide](http://babeljs.io/docs/setup/).
+
+Once you have Babel setup, install `babel-preset-flow` with either [Yarn](https://yarnpkg.com/) or [npm](https://www.npmjs.com/).
+
+```bash
+yarn add --dev babel-preset-flow
+# or
+npm install --save-dev bebel-preset-flow
```
Comment (yangshun) [inline content/docs/static-type-checking.md]: Typo: `babel-preset-flow`

File: content/docs/static-type-checking.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+[Create React App](https://github.com/facebookincubator/create-react-app) already supports Flow by default. Just [install Flow](https://flow.org/en/docs/install/) and create a `.flowconfig` file by running `flow init`.
+
+```bash
+create-react-app my-app && cd my-app
+yarn add --dev flow-bin
+yarn run flow init
+```
+
+Flow will be run as part of create-react-app’s scripts.
```
Comment (yangshun) [inline content/docs/static-type-checking.md]: `create-react-app` should be in code here.

File: content/docs/static-type-checking.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+## Using TypeScript with React
+
+For a fast dive into TypeScript, go [here](https://www.typescriptlang.org/play/).
+
+### With Create React App
+
+All you need to do is:
+
+```
```
Comment (yangshun) [inline content/docs/static-type-checking.md]: Would be great to add `bash/sh` highlighting here

File: content/docs/static-type-checking.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+create-react-app my-app --scripts-version=react-scripts-ts
+```
+
+[react-scripts-ts](https://www.npmjs.com/package/react-scripts-ts) is a set of adjustments to take the standard create-react-app project pipeline and bring TypeScript into the mix.
```
Comment (yangshun) [inline content/docs/static-type-checking.md]: `create-react-app`

File: content/docs/static-type-checking.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
+---
+id: static-type-checking
+title: Static Type Checking
+permalink: docs/static-type-checking.html
+prev: typechecking-with-prototypes.html
+next: refs-and-the-dom.html
+---
+
+As your app grows, you can catch and eliminate an entire category of bugs with static typechecking. You can use JavaScript extensions like [Flow](https://flowtype.org/) or [TypeScript](https://www.typescriptlang.org/) to statically typecheck your whole application. In complex applications, we recommend to use Flow or TypeScript for typechecking instead of typechecking with PropTypes.
```
Comment (yangshun) [inline content/docs/static-type-checking.md]: Should `type check` should be two words? The [Flow website](https://flow.org/en/docs/getting-started/) uses two words

**Comment by bvaughn** [comment]
Thank you for this PR. I'm sorry it took me so long to review it. I have a couple of minor grammatical/wordsmithing nits that I'm going to just make myself so that I can merge the PR now while I have the time (so that I don't get distracted again). Hope that's okay! 🙇

File: content/docs/static-type-checking.md (hunk 3 above)
Comment (gaearon) [inline content/docs/static-type-checking.md]: This is not true. There is no integration between CRA and Flow except that CRA allows Flow syntax.

CRA already has correct documentation for adding Flow: https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/template/README.md#adding-flow

We should probably just point people there 😉

File: content/docs/static-type-checking.md (hunk 5 above)
Comment (gaearon) [inline content/docs/static-type-checking.md]: This is not an officially supported project. We should at least clarify this.

File: content/docs/static-type-checking.md (hunk 1 above)
Comment (gaearon) [inline content/docs/static-type-checking.md]: Placing Babel first will cause CRA users to try to follow these instructions (and fail).

**Comment by gaearon** [comment]
Left a few comments for future reference. I'll edit this—just hoping it's useful for future readers. :-)

File: content/docs/static-type-checking.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+npm install --save-dev bebel-preset-flow
+```
+
+Then add `flow` to your Babel presets config.
+
+```
+{
+  "presets": ["flow"]
+}
+```
```
Comment (gaearon) [inline content/docs/static-type-checking.md]: This section only explains how to add Flow syntax support to a project using Babel. But it doesn't actually explain how to run Flow.

**Comment by bvaughn** [comment]
I feel like I kind of failed @ishankbahl with this PR, given all of the corrections Dan's making. I don't know much about `create-react-app`. Have never used it in depth. Was just trying to help keep PRs moving. \~shrug~

**Comment by gaearon** [comment]
That's the right way to do it—merge something that's a good first step and then iterate. :-)

I rewrote these parts in https://github.com/reactjs/reactjs.org/pull/303 but wouldn't have done it if you both didn't start the ball rolling!

**Pull request merged**